### PR TITLE
[WIP] Add a doc on how we work with feature gates

### LINKED ISSUE #5542: Document feature flag lifecycle
The API lifecycle is well documented, but the feature flag lifecycle, which is related but not strictly coupled to it is not. It would be great to document the conventions for alpha, beta, and GA feature flags, how to add and graduate them, what the conventions for locking to true mean, etc.

This issue sparked by the conversation in https://github.com/kubernetes/kubernetes/pull/99158#issuecomment-783597035

/sig architecture
/area developer-guide
/assign @dims

### PR DESCRIPTION
**Pull request by dims**
Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**Which issue(s) this PR fixes**:

Fixes #5542

### CODE REVIEW

File: contributors/devel/sig-architecture/feature-gates.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Feature gate progression in Kubernetes
+
+Feature flags are how Kubernetes tracks progression of functionality that is available to a deployer.
```
Comment (ehashman) [inline contributors/devel/sig-architecture/feature-gates.md:3]:
```suggestion
Feature flags are how Kubernetes tracks progression of functionality that is available when deployed.
```

File: contributors/devel/sig-architecture/feature-gates.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+* `PreRelease` is set to `featuregate.Alpha`
+* `Default` is always set to `false`
+* `LockToDefault` is not set. Defaults to `false`
+
+This enables folks to switch on the feature using the command line. By default it is not switched on.
+
+## Beta Features
+
+* `PreRelease` is set to `featuregate.Beta`
+* `Default` is always set to `true`
```
Comment (ehashman) [inline contributors/devel/sig-architecture/feature-gates.md:33]:
```suggestion
* `Default` is set to `true`
```

File: contributors/devel/sig-architecture/feature-gates.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## GA Features
+
+* `PreRelease` is set to `featuregate.GA`
+* `Default` is always set to `true`
+* `LockToDefault` is set to `true`
+
+GA features are always on. 
+
+Sometimes (rarely) we do not set `LockToDefault` and let it default to false. This enables folks to switch off the GA feature. We do this to tell folks that while the feature (for example `coredns`) is GA, they need to move off say `kubedns` to `coredns` in their infrastructures pretty soon and if they want to continue `kubedns` for a short time, they will have to switch off the GA flag. When we do remove the support for `kubedns` entirely we would set `LockToDefault` to `true` with some grace period for the transition
```
Comment (ehashman) [inline contributors/devel/sig-architecture/feature-gates.md:48]:
```suggestion
Sometimes (rarely) we do not set `LockToDefault` and let it default to `false`. This enables folks to switch off the GA feature. We do this to tell folks that while the feature (for example `coredns`) is GA, they need to move off say `kubedns` to `coredns` in their infrastructures pretty soon and if they want to continue `kubedns` for a short time, they will have to switch off the GA flag. When we do remove the support for `kubedns` entirely we would set `LockToDefault` to `true` with some grace period for the transition.
```

File: contributors/devel/sig-architecture/feature-gates.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+Also note that when we set `LockToDefault` to `true`, we remove all references (if/then conditions) to the feature gate from the codebase. 
+
+## Deprecation
+
+* `PreRelease` is set to `featuregate.Deprecated`
+* `Default` is always set to `true`
+
+## Other scenarios
+
+Sometimes we use`{Default: true, PreRelease: featuregate.Beta}` for keeping legacy behavior (see `LegacyNodeRoleBehavior`) on and when we basically want to drop the legacy behavior with new default behavior. So at GA we will end up with `{Default: false, PreRelease: featuregate.GA, LockToDefault: true}`
```
Comment (ehashman) [inline contributors/devel/sig-architecture/feature-gates.md:62]: I don't understand this comment, there might be a typo or we might want to elaborate more.

File: contributors/devel/sig-architecture/feature-gates.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+Typically we add a comment `// remove in 1.23` to signal when we would entirely remove the feature gate. Remember when the feature gate is removed and the deployer has forgotten to drop the reference to the feature in the CLI flags (say the `kube-apiserver`), then they will see a hard failure. 
+
+Also note that when we set `LockToDefault` to `true`, we remove all references (if/then conditions) to the feature gate from the codebase. 
+
+## Deprecation
+
+* `PreRelease` is set to `featuregate.Deprecated`
+* `Default` is always set to `true`
```
Comment (fabriziopandini) [inline contributors/devel/sig-architecture/feature-gates.md:58]: Should this depend on the current state (alpha/beta)?

File: contributors/devel/sig-architecture/feature-gates.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
+# Feature gate progression in Kubernetes
+
+Feature flags are how Kubernetes tracks progression of functionality that is available to a deployer.
+
+## Lifecycle
+
+Feature progress through `Alpha` -> `Beta` -> `GA`. Sometimes we end up deciding that a feature is not going to be supported and we end up marking them as `Deprecated`.
```
Comment (fabriziopandini) [inline contributors/devel/sig-architecture/feature-gates.md:7]: Might be worth indicating that there is also a final state when the feature gates disappear and the feature became part of the product

**Comment by fabriziopandini** [comment]
It will be great to link a reference implementation or even better a package providing support for feature gates across the board (meanings re-usable also in Kubernetes-sigs projects)

File: contributors/devel/sig-architecture/feature-gates.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+## Beta Features
+
+* `PreRelease` is set to `featuregate.Beta`
+* `Default` is always set to `true`
+* `LockToDefault` is not set. Defaults to `false`
+
+This enables the feature to be on and available in the default installation of kubernetes. 
+
+Sometimes (rarely) the `Default` is set to `false`. For example when this flag needs some explicit action or support outside of Kubernetes. This tells folks that while this feature is Beta, you still need to do some work to switch it on and use it.
```
Comment (nikhita) [inline contributors/devel/sig-architecture/feature-gates.md:38]: Would be great if we could add an example here. Maybe the CSIMigration APIs?

https://github.com/kubernetes/kubernetes/blob/4cc1127e9251fff364d5c77e2a9a9c3ad42383ab/pkg/features/kube_features.go#L699

File: contributors/devel/sig-architecture/feature-gates.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+* `LockToDefault` is set to `true`
+
+GA features are always on. 
+
+Sometimes (rarely) we do not set `LockToDefault` and let it default to false. This enables folks to switch off the GA feature. We do this to tell folks that while the feature (for example `coredns`) is GA, they need to move off say `kubedns` to `coredns` in their infrastructures pretty soon and if they want to continue `kubedns` for a short time, they will have to switch off the GA flag. When we do remove the support for `kubedns` entirely we would set `LockToDefault` to `true` with some grace period for the transition
+
+
+Typically we add a comment `// remove in 1.23` to signal when we would entirely remove the feature gate. Remember when the feature gate is removed and the deployer has forgotten to drop the reference to the feature in the CLI flags (say the `kube-apiserver`), then they will see a hard failure. 
+
+Also note that when we set `LockToDefault` to `true`, we remove all references (if/then conditions) to the feature gate from the codebase.
```
Comment (nikhita) [inline contributors/devel/sig-architecture/feature-gates.md:53]: can we add a note that `codebase` here refers to the `kube_features.go` files?

**Comment by nikhita** [comment]
> or even better a package providing support for feature gates across the board (meanings re-usable also in Kubernetes-sigs projects)

@fabriziopandini did you mean being able to import the feature gates in `kube_features.go` or being able to import/reuse the feature gate implementation in general?

If it's the latter, the implementation is defined in component-base, so it can be reused today - https://github.com/kubernetes/component-base/blob/master/featuregate/feature_gate.go

**Comment by fabriziopandini** [comment]
> If it's the latter, the implementation is defined in component-base, so it can be reused today - https://github.com/kubernetes/component-base/blob/master/featuregate/feature_gate.go

Great! let's link this in the doc!

File: contributors/devel/sig-architecture/feature-gates.md (hunk 7 above)
Comment (pacoxu) [reply contributors/devel/sig-architecture/feature-gates.md:38]: I suggest adding more comments if it is locked to false. It is really weried for beginners.

File: contributors/devel/sig-architecture/feature-gates.md (hunk 4 above)
Comment (pacoxu) [reply contributors/devel/sig-architecture/feature-gates.md:62]: LockToDefault=true， means the default value is false and cannot be set to true here.

Only this case: https://github.com/kubernetes/kubernetes/pull/97543 as designed.

**Comment by dims** [comment]
cc @kikisdeliveryservice

**Comment by dims** [comment]
closing in favor of https://github.com/kubernetes/community/pull/5630 thanks @kikisdeliveryservice

**Comment by nikhita** [comment]
> closing in favor of #5630 thanks @kikisdeliveryservice

Closing this for real. :stuck_out_tongue: 

Thanks for picking this up, @kikisdeliveryservice!

**Pull request closed without merging**